src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,729 @@
+        {
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<(TElement, TPriority)>())
+            {
+                // Clear the elements so that the gc can reclaim the references
+                Array.Clear(_nodes, 0, _size);
+            }
+            _size = 0;
+            _version++;
+        }
+
+        /// <summary>
+        /// Ensures that the priority queue has the specified capacity
+        /// and resizes its underlying array if necessary.
+        /// </summary>
+        public int EnsureCapacity(int capacity)
+        {
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(capacity), capacity, SR.ArgumentOutOfRange_NeedNonNegNum);
+            }
+
+            int currentCapacity = _nodes.Length;
+            if (currentCapacity >= capacity)
+            {
+                return currentCapacity;
+            }
+
+            int capacityForNextGrowth = ComputeCapacityForNextGrowth();
+            if (capacityForNextGrowth > capacity)
+            {
+                capacity = capacityForNextGrowth;
+            }
+
+            SetCapacity(capacity);
+            return capacity;
+        }
+
+        /// <summary>
+        /// Sets the capacity to the actual number of items in the priority queue,
+        /// if that is less than 90 percent of current capacity.
+        /// </summary>
+        public void TrimExcess()
+        {
+            int threshold = (int)(_nodes.Length * 0.9);
+            if (_size < threshold)
+            {
+                SetCapacity(_size);
+            }
+        }
+
+        private void EnsureEnoughCapacityBeforeAddingNode()
+        {
+            if (_size < _nodes.Length)
+            {
+                return;
+            }
+
+            int newCapacity = ComputeCapacityForNextGrowth();
+            SetCapacity(newCapacity);
+        }
+
+        private int ComputeCapacityForNextGrowth()
+        {
+            const int GrowthFactor = 2;
+            const int MaxArrayLength = 0X7FEFFFFF;
+
+            int newCapacity = _nodes.Length * GrowthFactor;
+            if (newCapacity < _nodes.Length + MinimumElementsToGrowBy)
+            {
+                newCapacity = _nodes.Length + MinimumElementsToGrowBy;
+            }
+
+            // Allow the structure to grow to maximum possible capacity (~2G elements) before encountering overflow.
+            // Note that this check works even when _nodes.Length overflowed thanks to the (uint) cast.
+
+            if ((uint)newCapacity > MaxArrayLength)
+            {
+                newCapacity = MaxArrayLength;
+            }
+
+            return newCapacity;
+        }
+
+        /// <summary>
+        /// Grows or shrinks the array holding nodes. Capacity must be >= _size.
+        /// </summary>
+        private void SetCapacity(int capacity)
+        {
+            Array.Resize(ref _nodes, capacity);
+            _version++;
```

Comment:
Why would we need to increment `_version` here? Resizing the buffer should not on its own invalidate existing enumerators. On the other hand, most methods calling `SetCapacity` will already increment `_version` if required.

Reply:
That's the existing behavior in `Queue.SetCapacity`, excerpt from there:

```
private void SetCapacity(int capacity)
{
    T[] newarray = new T[capacity];
    if (_size > 0)
    {
        if (_head < _tail)
        {
            Array.Copy(_array, _head, newarray, 0, _size);
        }
        else
        {
            Array.Copy(_array, _head, newarray, 0, _array.Length - _head);
            Array.Copy(_array, 0, newarray, _array.Length - _head, _tail);
        }
    }

    _array = newarray;
    _head = 0;
    _tail = (_size == capacity) ? 0 : _size;
    _version++;
}
```

This method (and consequently `_version++`) impacts public methids `TrimExcess`, `Enqueue` in `Queue`.

Reply:
In `Dictionary.Generic.Tests.cs`, it's tested that `EnsureCapacity` and `TrimExcess` invalidate enumeration.